libclamav_rust/src/fuzzy_hash.rs

```diff
@@ -110,212 +118,205 @@ pub struct FuzzyHashMeta {
-
-            let fuzzy_hash = FuzzyHash::Image(image_fuzzy_hash);
-
-            let meta: FuzzyHashMeta = FuzzyHashMeta {
-                lsigid: lsig_id,
-                subsigid: subsig_id,
-                #[cfg(feature = "not_ready")]
-                hamming_distance: distance,
-            };
-
-            // If the hash key does not exist in the hashmap, insert an empty vec.
-            // Then add the current meta struct to the entry.
-            hashmap
-                .entry(fuzzy_hash)
-                .or_insert_with(Vec::new)
-                .push(meta);
-
-            Ok(())
-        }
-        _ => {
-            error!("Unknown fuzzy hash algorithm: {}", algorithm);
-            Err(FuzzyHashError::UnknownAlgorithm(algorithm.to_string()))
-        }
-    }
+    frs_call!(
+        err = err,
+        hashmap.load_subsignature(hexsig, lsig_id, subsig_id)
+    )
 }
 
 /// C interface for fuzzy_hash_calculate_image().
 /// Handles all the unsafe ffi stuff.
+///
+/// # Safety
+///
+/// `file_bytes` and `hash_out` must not be NULL
 #[export_name = "fuzzy_hash_calculate_image"]
-pub extern "C" fn _fuzzy_hash_calculate_image(
+pub unsafe extern "C" fn _fuzzy_hash_calculate_image(
     file_bytes: *const u8,
     file_size: usize,
     hash_out: *mut u8,
     hash_out_len: usize,
     err: *mut *mut FRSError,
 ) -> bool {
-    if file_bytes.is_null() {
-        error!("invalid NULL pointer for image input buffer");
-        return false;
-    }
     if hash_out.is_null() {
-        error!("invalid NULL pointer for hash output buffer");
-        return false;
+        return frs_error!(err = err, FuzzyHashError::NullParam("hash_out"));
     }
 
-    let buffer: &[u8] = unsafe { slice::from_raw_parts(file_bytes, file_size) };
+    let buffer = if file_bytes.is_null() {
+        return frs_error!(err = err, FuzzyHashError::NullParam("file_bytes"));
+    } else {
+        slice::from_raw_parts(file_bytes, file_size)
+    };
 
     let hash_result = fuzzy_hash_calculate_image(buffer);
     let hash_bytes = match hash_result {
         Ok(hash) => hash,
-        Err(error) => {
-            error!("Failed to calculate image fuzzy hash: {}", error);
-            let hash_result: Result<(), FuzzyHashError> = Err(error);
-            return unsafe { frs_result!(result_in = hash_result, err = err) };
-        }
+        Err(error) => return frs_error!(err = err, error),
     };
 
     if hash_out_len < hash_bytes.len() {
-        let hash_result: Result<(), FuzzyHashError> =
-            Err(FuzzyHashError::InvalidParameter(format!(
+        return frs_error!(
+            err = err,
+            FuzzyHashError::InvalidParameter(format!(
                 "hash_bytes output parameter too small to hold the hash: {} < {}",
                 hash_out_len,
                 hash_bytes.len()
-            )));
-        return unsafe { frs_result!(result_in = hash_result, err = err) };
+            ))
+        );
```

Comment:
This also feels much better not having to construct a Result and then consume it.  Very nice.